spec.md

```diff
@@ -547,212 +547,134 @@ See [@sec:mem] for more details.
-Reference types may not contain other reference types.
-Property types represent information about the circuit that is not hardware.
+FIRRTL has four classes of types: **ground** types, **aggregate** types, **probe** types, and **property** types.
 
 ## Ground Types
 
-There are five classes of ground types in FIRRTL: unsigned integer types, signed integer types, a clock type, reset types, and analog types.
+Ground types are types which are not composed of other types.
+They are the integer types, clocks, and resets.
 
 ### Integer Types
 
-Both unsigned and signed integer types may optionally be given a known non-negative integer bit width.
+FIRRTL supports both signed and unsigned integer types.
 
-``` firrtl
-UInt ; unsigned int type with inferred width
-SInt ; signed int type with inferred width
-UInt<10> ; unsigned int type with 10 bits
-SInt<32> ; signed int type with 32 bits
-```
-
-Alternatively, if the bit width is omitted, it will be automatically inferred by FIRRTL's width inferencer, as detailed in [@sec:width-inference].
-
-#### Zero Bit Width Integers
-
-Integers of width zero are permissible.
-They are always zero extended.
-Thus, when used in an operation that extends to a positive bit width, they behave like a zero.
-While zero bit width integers carry no information, we allow 0-bit integer constant zeros for convenience: `UInt<0>(0)`{.firrtl} and `SInt<0>(0)`{.firrtl}.
-
-``` firrtl
-wire zero_u : UInt<0>
-invalidate zero_u
-wire zero_s : SInt<0>
-invalidate zero_s
-
-wire one_u : UInt<1>
-connect one_u, zero_u
-wire one_s : SInt<1>
-connect one_s, zero_s
-```
-
-Is equivalent to:
-
-``` firrtl
-wire one_u : UInt<1>
-connect one_u, UInt<1>(0)
-wire one_s : SInt<1>
-connect one_s, SInt<1>(0)
-```
-
-### Clock Type
-
-The clock type is used to describe wires and ports meant for carrying clock signals.
-The usage of components with clock types are restricted.
-Clock signals cannot be used in most primitive operations, and clock signals can only be connected to components that have been declared with the clock type.
+`UInt<n>`{.firrtl} is an `n`-bit wide unsigned integer.
+`SInt<n>`{.firrtl} is an `n`-bit wide signed integer.
 
-The clock type is specified as follows:
-
-``` firrtl
-Clock
-```
-
-### Reset Type
-
-The uninferred `Reset`{.firrtl} type is either inferred to `UInt<1>`{.firrtl} (synchronous reset) or `AsyncReset`{.firrtl} (asynchronous reset) during compilation.
-
-``` firrtl
-Reset ; inferred type
-AsyncReset
+```firrtl
+UInt<10> ; a 10-bit unsigned integer
+SInt<32> ; 32-bit signed integer
 ```
 
-Synchronous resets used in registers will be mapped to a hardware description language representation for synchronous resets.
+Both `UInt<0>`{.firrtl} and `SInt<0>`{.firrtl} are valid types.
+They are considered to have only a single value representing zero (written as either `UInt<0>(0)`{.firrtl} or `SInt<0>(0)`{.firrtl}).
+They behave like zero when they are extended to a positive width integer type.
```

Comment:
Using `pad` with a reference to the `pad` primop may be more accurate than saying `extend` which could mean either zero-extend or sign-extend (and FIRRTL doesn't have both operations, just one operation that is sensitive to the type).

As commentary (either in the PR or future), interpreting sign-extension of a zero-width signed integer as adding leading zeros was something controversial when I tried to land this on upstream MLIR. There were discussions about whether this should _actually add ones_. I.e., the value of a zero-width signed integer is actually `-1`. The end result was that this was undefined and MLIR will error if you try to sign extend a zero-width signed integer.

This is again in the Bruce Hornsby bucket and it's fine that FIRRTL does it this way. A reader with experience (or a mathematical background) may do a double-take on these sentences.

Did you have a similar reaction given your math background?

Reply:
This section is the original copy.

I'll weigh in on the `SInt<0>` battle at some point. But not today :)